components/layout/query.rs

```diff
@@ -44,9 +44,94 @@ use crate::flow::inline::construct::{TextTransformation, WhitespaceCollapse};
+/// mapping of coordinates, scroll offset, etc.
+pub(crate) struct PostCompositeQueryHelper<'dom> {
+    /// Get the scroll offset of a viewport (i.e. window).
+    pub root_scroll_offset_getter: &'dom dyn Fn() -> Vector2D<Au>,
+    /// Get the scroll offset of a BoxFragment.
+    pub scroll_offset_getter: &'dom dyn Fn(&BoxFragment) -> Vector2D<Au>,
+}
+
+impl PostCompositeQueryHelper<'_> {
+    fn retrieve_box_fragment(fragment: &Fragment) -> Option<&ArcRefCell<BoxFragment>> {
+        match fragment {
+            Fragment::Box(box_fragment) | Fragment::Float(box_fragment) => Some(box_fragment),
+            _ => None,
+        }
+    }
+
+    /// Get the position property for current node.
+    fn retrieve_position_property<'dom, Node: LayoutNode<'dom> + 'dom>(
+        node: Node,
+    ) -> Option<Position> {
+        node.fragments_for_pseudo(None)
+            .first()
+            .and_then(PostCompositeQueryHelper::retrieve_box_fragment)
+            .map(|fragment| fragment.borrow().style.get_box().position)
+    }
+
+    /// Find the parent node in containing block chain.
+    fn parent_containing_block<'dom, Node: LayoutNode<'dom> + 'dom>(
+        node: Node,
+        position: Option<Position>,
+    ) -> Option<Node> {
+        fn established_containing_block(
+            style: &ServoArc<ComputedValues>,
+            fragment_flags: FragmentFlags,
+            position: Option<Position>,
+        ) -> bool {
+            match position {
+                Some(Position::Absolute) => {
+                    style.establishes_containing_block_for_absolute_descendants(fragment_flags)
+                },
+                Some(Position::Fixed) => {
+                    style.establishes_containing_block_for_all_descendants(fragment_flags)
+                },
+                _ => true,
+            }
+        }
+
+        let mut maybe_parent_node = node.parent_node();
+        while let Some(parent_node) = maybe_parent_node {
+            if let Some(parent_fragment) = parent_node.fragments_for_pseudo(None).first() {
+                let established_containing_block = match parent_fragment {
+                    Fragment::Box(box_fragment) | Fragment::Float(box_fragment) => {
+                        established_containing_block(
+                            &box_fragment.borrow().style,
+                            box_fragment.borrow().base.flags,
+                            position,
+                        )
+                    },
+                    Fragment::Positioning(fragment) => {
+                        fragment.borrow().style.as_ref().is_some_and(|style| {
+                            established_containing_block(
+                                style,
+                                fragment.borrow().base.flags,
+                                position,
+                            )
+                        })
+                    },
+                    _ => continue,
```

Comment:
Isn't this `continue` going to create a dead loop?

Reply:
Fixed! Thanks for noticing it!